### PR TITLE
add `preferedType` to weapon roll options

### PR DESCRIPTION
**Pull request by amir-arad**
this will allow macros to fully automate type of attacks, in addition to `skipDialog`

another API option I would like to add if this is acceptable: attack modifier

### CODE REVIEW

File: src/module/item/entity.js [hunk 1]
```diff
@@ -84,29 +84,33 @@ export class OseItem extends Item {
...
-      new Dialog({
-        title: "Choose Attack Range",
-        content: "",
-        buttons: {
-          melee: {
-            icon: '<i class="fas fa-fist-raised"></i>',
-            label: "Melee",
-            callback: () => {
-              this.actor.targetAttack(rollData, "melee", options);
+      if (options.preferedType === "melee" || options.preferedType === "missile"){
```
Comment (wyrmisis) [inline src/module/item/entity.js:87]: I wonder if there's a way we could express this without nested if statements?

File: src/module/item/entity.js [hunk 2]
```diff
@@ -84,29 +84,33 @@ export class OseItem extends Item {
...
+        type = options.preferedType;
+      } else {
+        // Dialog
+        new Dialog({
+          title: "Choose Attack Range",
```
Comment (wyrmisis) [inline src/module/item/entity.js:92]: While you're in here, could you replace strings like this with a localized string?

**Review by wyrmisis — CHANGES_REQUESTED** [review]

File: src/module/item/entity.js (hunk 1 above)
Comment (amir-arad) [reply src/module/item/entity.js]: done

File: src/module/item/entity.js (hunk 2 above)
Comment (amir-arad) [reply src/module/item/entity.js]: done

File: src/module/item/entity.js [hunk 3]
```diff
@@ -84,29 +84,33 @@ export class OseItem extends Item {
...
-          melee: {
-            icon: '<i class="fas fa-fist-raised"></i>',
-            label: "Melee",
-            callback: () => {
-              this.actor.targetAttack(rollData, "melee", options);
+      if (options.preferedType === "melee" || options.preferedType === "missile"){
+        type = options.preferedType;
+      } else {
+        // Dialog
+        new Dialog({
```
[the remark this answers is not in the document]
Comment (amir-arad) [reply src/module/item/entity.js:87]: done (+poetic license)

File: src/module/item/entity.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (amir-arad) [reply src/module/item/entity.js]: done

File: src/module/item/entity.js [hunk 4]
```diff
@@ -68,49 +68,48 @@ export class OseItem extends Item {
...
+          default: "melee",
+        }).render(true);
+      });
+    } 
+    if (itemData.missile) return CONFIG.OSE.attack_types.missile;
+    return CONFIG.OSE.attack_types.melee;
+  }
+
+  rollWeapon(options = {}) {
+    this.getAttackType()
```
Comment (anthonyronda) [inline src/module/item/entity.js:111]: options not passed

rolling an attack fails with

```diff
- Uncaught (in promise) TypeError: can't access property "preferedType", options is undefined
-    getAttackType entity.js:75
-    rollWeapon entity.js:111
-    _rollAbility actor-sheet.js:171
-    activateListeners actor-sheet.js:504
```

File: src/module/config.ts [hunk 5]
```diff
@@ -49,6 +50,7 @@ export type Color =
   | "blue"
   | "orange"
   | "white";
+export type AttackType = "missile" | "melee" | "attack";
```
Comment (anthonyronda) [inline src/module/config.ts:53]: possibly merge from main first and rewrite using the method you created in #369 ?

File: src/module/item/entity.js [hunk 6]
```diff
@@ -68,49 +68,48 @@ export class OseItem extends Item {
...
-        buttons: {
-          melee: {
-            icon: '<i class="fas fa-fist-raised"></i>',
-            label: "Melee",
-            callback: () => {
-              this.actor.targetAttack(rollData, "melee", options);
-            },
+    if (itemData.missile && itemData.melee) {
+      if (CONFIG.OSE.attack_types[options.preferedType]){
+        return CONFIG.OSE.attack_types[options.preferedType];
```
Comment (anthonyronda) [inline src/module/item/entity.js:76]: preferredType, extra r

Spelling is hard! I wouldn't mention it if there wasn't something else to change

**Review by anthonyronda — CHANGES_REQUESTED** [review]
hi! we have a problem, please see below

let me know if you don't have time for this in the next 24 hours

**Comment by amir-arad** [comment]
closing this PR, intend to re-open from main of next version, after I also learn to test locally...

**Comment by anthonyronda** [comment]
I will write some documentation about writing tests today

**Pull request closed without merging**